Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -46,16 +44,36 @@
 {
     constexpr uint32_t maximumMessageSize { std::numeric_limits<uint32_t>::max() };
     constexpr bool rejectOversizedMessages { false };
-    nw_connection_group_set_receive_handler(connectionGroup, maximumMessageSize, rejectOversizedMessages, makeBlockPtr([weakThis = WeakPtr { *this }] (dispatch_data_t, nw_content_context_t, bool) {
-        // FIXME: Implement.
+    nw_connection_group_set_receive_handler(connectionGroup, maximumMessageSize, rejectOversizedMessages, makeBlockPtr([weakThis = WeakPtr { *this }] (dispatch_data_t datagram, nw_content_context_t, bool) {
+        RefPtr strongThis = weakThis.get();
+        if (!strongThis)
+            return;
+
+        // FIXME: Not only is this an unnecessary string copy, but it's also something that should probably be in WTF or FragmentedSharedBuffer.
+        auto vectorFromData = [](dispatch_data_t content) {
+            ASSERT(content);
+            __block Vector<uint8_t> request;
+            dispatch_data_apply(content, ^bool(dispatch_data_t, size_t, const void* buffer, size_t size) {
+                request.append(std::span { static_cast<const uint8_t*>(buffer), size });
+                return true;
+            });
+            return request;
+        };
+
+        strongThis->receiveDatagram(vectorFromData(datagram).span());
     }).get());
 
     // FIXME: Use nw_connection_group_set_new_connection_handler to receive incoming connections.
 }
 
-void NetworkTransportSession::initialize(NetworkConnectionToWebProcess& connectionToWebProcess, URL&& url, CompletionHandler<void(std::unique_ptr<NetworkTransportSession>&&)>&& completionHandler)
+void NetworkTransportSession::initialize(NetworkConnectionToWebProcess& connectionToWebProcess, URL&& url, CompletionHandler<void(RefPtr<NetworkTransportSession>&&)>&& completionHandler)
 {
-    RetainPtr endpoint = adoptNS(nw_endpoint_create_url(url.string().utf8().data()));
+    // FIXME: Use nw_endpoint_create_url to support all URLs on systems where rdar://135036170 is fixed.
+    bool isIPv4Address = URL::hostIsIPAddress(url.host()) && !url.host().contains(':');
```

Comment:
More of a general comment: seems like we should provide a variant of `hostIsIPAddress` that returns whether it's an IPv4 or v6 address.

Reply:
Not a bad idea.

Reply:
It looks like nw_endpoint_create_host accepts more than just IP addresses.  I'll try removing this restriction.